// @maxmind/geoip2-node/dist/src/models/Country.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const camelcaseKeys = require("camelcase-keys");
class Country {
    constructor(response) {
        const camelcaseResponse = camelcaseKeys(response, {
            deep: true,
            exclude: [/\-/],
        });
        this.continent = camelcaseResponse.continent || {};
        this.country = camelcaseResponse.country || {};
        this.maxmind = camelcaseResponse.maxmind || {};
        this.registeredCountry =
            this.setBooleanRegisteredCountry(camelcaseResponse.registeredCountry) ||
                {};
        this.representedCountry = camelcaseResponse.representedCountry || {};
        this.traits = this.setBooleanTraits(camelcaseResponse.traits || {});
    }
    setBooleanTraits(traits) {
        const booleanTraits = [
            'isAnonymous',
            'isAnonymousProxy',
            'isAnonymousVpn',
            'isHostingProvider',
            'isLegitimateProxy',
            'isPublicProxy',
            'isSatelliteProvider',
            'isTorExitNode',
        ];
        booleanTraits.forEach(trait => {
            traits[trait] = !!traits[trait];
        });
        return traits;
    }
    setBooleanRegisteredCountry(country) {
        if (country) {
            country.isInEuropeanUnion = !!country.isInEuropeanUnion;
        }
        return country;
    }
}
exports.default = Country;
