diff --git a/src/bitcoind.cpp b/src/bitcoind.cpp
index 17989a421..8ca05951d 100644
--- a/src/bitcoind.cpp
+++ b/src/bitcoind.cpp
@@ -112,11 +112,12 @@ static bool AppInit(int argc, char* argv[])
             // InitError will have been called with detailed error, which ends up on console
             return false;
         }
-        if (!AppInitSanityChecks())
-        {
-            // InitError will have been called with detailed error, which ends up on console
-            return false;
-        }
+        // if (!AppInitSanityChecks())
+        // {
+        //     // InitError will have been called with detailed error, which ends up on console
+        //     return false;
+        // }
+        LogPrintf("ilios:After SanityChecks\n");
         if (gArgs.GetBoolArg("-daemon", false))
         {
 #if HAVE_DECL_DAEMON
